[Crash] Crash in instrumented tests: java.lang.RuntimeException: Failed to call observer method

[comment]: # (Please do your best to search for duplicate issues before filing a new issue so we can keep our issue board clean)
[comment]: # (Every issue should have the exact bug and steps to reproduce described in it. Please do not file feedback list tickets as it is difficult to parse them and address their individual points)
[comment]: # (Read https://github.com/mozilla-mobile/fenix#i-want-to-file-an-issue for more information)

## Steps to reproduce
Failing UI test: org.mozilla.fenix.ui.NavigationToolbarTest#goBackTest

Test report: https://console.firebase.google.com/u/0/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/8229397304259493389/executions/bs.419e48010fba6b1a

[Test report logcat ](https://github.com/mozilla-mobile/fenix/files/5696223/logcat.txt)


StackTrace:
`java.lang.RuntimeException: Failed to call observer method
     FATAL EXCEPTION: DefaultDispatcher-worker-3
Process: org.mozilla.fenix.debug, PID: 12996
java.lang.RuntimeException: Failed to call observer method
	at androidx.lifecycle.ClassesInfoCache$MethodReference.invokeCallback(ClassesInfoCache.java:226)
	at androidx.lifecycle.ClassesInfoCache$CallbackInfo.invokeMethodsForEvent(ClassesInfoCache.java:194)
	at androidx.lifecycle.ClassesInfoCache$CallbackInfo.invokeCallbacks(ClassesInfoCache.java:185)
	at androidx.lifecycle.ReflectiveGenericLifecycleObserver.onStateChanged(ReflectiveGenericLifecycleObserver.java:37)
	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.java:361)
	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.java:300)
	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.java:339)
	at androidx.lifecycle.LifecycleRegistry.addObserver(LifecycleRegistry.java:196)
	at mozilla.components.lib.state.ext.StoreExtensionsKt.flow(StoreExtensions.kt:155)
	at mozilla.components.lib.state.ext.FragmentKt.consumeFlow(Fragment.kt:94)
	at mozilla.components.lib.state.ext.FragmentKt.consumeFlow$default(Fragment.kt:84)
	at org.mozilla.fenix.browser.BaseBrowserFragment.observeTabSource$app_debug(BaseBrowserFragment.kt:894)
	at org.mozilla.fenix.browser.BaseBrowserFragment$onViewCreated$1.invokeSuspend(BaseBrowserFragment.kt:235)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
Caused by: java.lang.IllegalThreadStateException: Must have a Handler
	at org.mozilla.geckoview.GeckoResult.then(GeckoResult.java:476)
	at org.mozilla.geckoview.GeckoResult.then(GeckoResult.java:366)
	at org.mozilla.geckoview.WebExtensionController.ensureBuiltIn(WebExtensionController.java:593)
	at mozilla.components.browser.engine.gecko.GeckoEngine.installWebExtension(GeckoEngine.kt:211)
	at mozilla.components.support.webextensions.WebExtensionController.install(WebExtensionController.kt:54)
	at mozilla.components.support.webextensions.WebExtensionController.install$default(WebExtensionController.kt:50)
	at mozilla.components.feature.accounts.FxaWebChannelFeature.start(FxaWebChannelFeature.kt:81)
	at mozilla.components.support.base.feature.ViewBoundFeatureWrapper.start$support_base_release(ViewBoundFeatureWrapper.kt:160)
	at mozilla.components.support.base.feature.LifecycleBinding.start(ViewBoundFeatureWrapper.kt:196)
	at java.lang.reflect.Method.invoke(Native Method)
	at androidx.lifecycle.ClassesInfoCache$MethodReference.invokeCallback(ClassesInfoCache.java:216)
	... 18 more`
### Expected behavior

### Actual behavior

### Device information

* Android device: Pixel 2, API 28, virtual device
* Fenix version: Debug 12/15

--- a/app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt
+++ b/app/src/main/java/org/mozilla/fenix/browser/BaseBrowserFragment.kt
@@ -230,10 +230,8 @@ abstract class BaseBrowserFragment : Fragment(), UserInteractionHandler,
 
         observeTabSelection(requireComponents.core.store)
 
-        lifecycleScope.launch(IO) {
-            if (!onboarding.userHasBeenOnboarded()) {
-                observeTabSource(requireComponents.core.store)
-            }
+        if (!onboarding.userHasBeenOnboarded()) {
+            observeTabSource(requireComponents.core.store)
         }
 
         requireContext().accessibilityManager.addAccessibilityStateChangeListener(this)

--- a/app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt
+++ b/app/src/main/java/org/mozilla/fenix/onboarding/FenixOnboarding.kt
@@ -6,6 +6,7 @@ package org.mozilla.fenix.onboarding
 
 import android.content.Context
 import android.content.SharedPreferences
+import android.os.StrictMode
 import androidx.annotation.VisibleForTesting
 import mozilla.components.support.ktx.android.content.PreferencesHolder
 import mozilla.components.support.ktx.android.content.intPreference
@@ -15,10 +16,13 @@ import org.mozilla.fenix.ext.components
 class FenixOnboarding(context: Context) : PreferencesHolder {
 
     private val metrics = context.components.analytics.metrics
-    override val preferences: SharedPreferences = context.getSharedPreferences(
-        PREF_NAME_ONBOARDING_KEY,
-        Context.MODE_PRIVATE
-    )
+    private val strictMode = context.components.strictMode
+    override val preferences: SharedPreferences = strictMode.resetAfter(StrictMode.allowThreadDiskReads()) {
+        context.getSharedPreferences(
+            PREF_NAME_ONBOARDING_KEY,
+            Context.MODE_PRIVATE
+        )
+    }
 
     private var onboardedVersion by intPreference(LAST_VERSION_ONBOARDING_KEY, default = 0)
 
@@ -27,7 +31,11 @@ class FenixOnboarding(context: Context) : PreferencesHolder {
         metrics.track(Event.DismissedOnboarding)
     }
 
-    fun userHasBeenOnboarded() = onboardedVersion == CURRENT_ONBOARDING_VERSION
+    fun userHasBeenOnboarded(): Boolean {
+        return strictMode.resetAfter(StrictMode.allowThreadDiskReads()) {
+            onboardedVersion == CURRENT_ONBOARDING_VERSION
+        }
+    }
 
     companion object {
         /**

--- a/app/src/test/java/org/mozilla/fenix/onboarding/FenixOnboardingTest.kt
+++ b/app/src/test/java/org/mozilla/fenix/onboarding/FenixOnboardingTest.kt
@@ -14,12 +14,17 @@ import org.junit.Assert.assertFalse
 import org.junit.Assert.assertTrue
 import org.junit.Before
 import org.junit.Test
+import org.junit.runner.RunWith
 import org.mozilla.fenix.components.metrics.Event
 import org.mozilla.fenix.components.metrics.MetricController
 import org.mozilla.fenix.ext.components
+import org.mozilla.fenix.helpers.FenixRobolectricTestRunner
+import org.mozilla.fenix.helpers.perf.TestStrictModeManager
 import org.mozilla.fenix.onboarding.FenixOnboarding.Companion.CURRENT_ONBOARDING_VERSION
 import org.mozilla.fenix.onboarding.FenixOnboarding.Companion.LAST_VERSION_ONBOARDING_KEY
+import org.mozilla.fenix.perf.StrictModeManager
 
+@RunWith(FenixRobolectricTestRunner::class)
 class FenixOnboardingTest {
 
     private lateinit var onboarding: FenixOnboarding
@@ -36,6 +41,7 @@ class FenixOnboardingTest {
         every { preferences.edit() } returns preferencesEditor
         every { metrics.track(any()) } returns Unit
         every { context.components.analytics.metrics } returns metrics
+        every { context.components.strictMode } returns TestStrictModeManager() as StrictModeManager
         every { context.getSharedPreferences(any(), MODE_PRIVATE) } returns preferences
 
         onboarding = FenixOnboarding(context)

